src/Dialect/ONNX/ONNXOps/Math/MatMul.cpp

```diff
@@ -189,6 +189,100 @@ LogicalResult ONNXMatMulIntegerOp::inferShapes(
   return shapeHelper.computeShapeAndUpdateType(elementType);
 }
 
+LogicalResult ONNXMatMulIntegerOp::verify() {
+  ONNXMatMulIntegerOp::Adaptor operandAdaptor(*this);
+  if (!hasShapeAndRank(getOperation()))
+    return success();
+
+  // If AZeroPoint is [M] (M != 1), A must be [M, K]
+  // If AZeroPoint rank is > 1, A must have the same rank, e.g.
+  // - [D1, D2, ..., DN, M, 1] if A is [D1, D2, ..., DN, M, K]
+  Value A = operandAdaptor.getA();
+  Value aZeroPoint = this->getAZeroPoint();
+  if (!isNoneValue(aZeroPoint)) {
+    auto aType = dyn_cast<ShapedType>(A.getType());
+    auto aZeroPointType = dyn_cast<ShapedType>(aZeroPoint.getType());
```

Comment:
since you rely on aType and aZeroPointType being non-null below, it's clearer to use `cast` instead of `dyn_cast`: both clearer to read the intention and also gives a clearer cast failure at runtime

alternatively, you can omit the explicit cast:
```
ShapedType aType, aZeroPointType = A.getType(), aZeroPointType.getType();
```
because ShapedType is a type interface: you can assign any Type to it and the constructor/assignment operator will check that the type is a ShapedType
https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Support/InterfaceSupport.h#L93-L99

Reply:
Thanks! Now I learned a new thing in MLIR.